code/modules/research/machinery/_production.dm

```diff
@@ -47,17 +59,160 @@
 		var/datum/design/d = SSresearch.techweb_design_by_id(i)
 		if((isnull(allowed_department_flags) || (d.departmental_flags & allowed_department_flags)) && (d.build_type & allowed_buildtypes))
 			cached_designs |= d
+	update_viewer_statics()
 
 /obj/machinery/rnd/production/RefreshParts()
 	calculate_efficiency()
+	ui_update()
+
+/obj/machinery/rnd/production/ui_interact(mob/user, datum/tgui/ui)
+	ui = SStgui.try_update_ui(user, src, ui)
+	if(!ui)
+		ui = new(user, src, "TechFab")
+		ui.open()
+		viewing_mobs += user
 
-/obj/machinery/rnd/production/ui_interact(mob/user)
-	if(!consoleless_interface)
-		return ..()
-	user.set_machine(src)
-	var/datum/browser/popup = new(user, "rndconsole", name, 460, 550)
-	popup.set_content(generate_ui())
-	popup.open()
+/obj/machinery/rnd/production/ui_close(mob/user)
+	. = ..()
+	viewing_mobs -= user
+
+/obj/machinery/rnd/production/proc/update_viewer_statics()
+	for(var/mob/M in viewing_mobs)
```

Comment:
```suggestion
	for(var/mob/M as() in viewing_mobs)
```

Reply:
Will address properly later, but I want to note this code is copied directly from modular fabricators, so suggestion probably applies there as well.